src/lando/main/scm/git.py

```diff
@@ -0,0 +1,258 @@
+import logging
+import os
+import subprocess
+import tempfile
+from contextlib import contextmanager
+from pathlib import Path
+from typing import ContextManager, Optional
+
+from lando.main.scm.exceptions import SCMException
+
+from .abstract_scm import AbstractSCM
+
+logger = logging.getLogger(__name__)
+
+ENV_COMMITTER_NAME = "GIT_COMMITTER_NAME"
+ENV_COMMITTER_EMAIL = "GIT_COMMITTER_EMAIL"
+
+
+class GitSCM(AbstractSCM):
+    """An implementation of the AbstractVCS for Git, for use by the Repo and LandingWorkers."""
+
+    DEFAULT_ENV = {
+        "GIT_SSH_COMMAND": (
+            'ssh -o "StrictHostKeyChecking no" -o "PasswordAuthentication no"'
+        ),
+    }
+
+    default_branch: str
+
+    def __init__(self, path: str, default_branch: str = "main"):
+        self.default_branch = default_branch
+        super().__init__(path)
+
+    def clone(self, source):
+        """Clone a repository from a source.
+        Args:
+            source: The source to clone the repository from.
+        Returns:
+            None
+        """
+        # When cloning, self.path doesn't exist yet, so we need to use another CWD.
+        self._git_run("clone", source, self.path, cwd="/")
+
+    def push(
+        self, push_path: str, target: Optional[str] = None, force_push: bool = False
+    ):
+        """Push local code to the remote repository.
+
+        Parameters:
+            push_path (str): The path to the repository where changes will be pushed.
+            target (Optional[str]): The target branch or reference to push to. Defaults to None.
+            force_push (bool): If True, force the push even if it results in a non-fast-forward update. Defaults to False.
+
+        Returns:
+            None
+        """
+        command = ["push"]
+        if force_push:
+            command += ["--force"]
+        command += [push_path]
+        if target:
+            command += [f"HEAD:{target}"]
```

Comment:
nit: I think it could be a tad bit easier to read if we sacrifice some repetition for readability, but not really sure.

Reply:
We could have a more hard-coded `command`, but with two `if`s, we'd end up with 3 or 4 branches, so something like

```python

if force_push && target:
  command = [ "push", "--force", f"HEAD:{target}" ]
elif force_push:
  command = [ "push", "--force" ]
elif target:
  command = [ "push", f"HEAD:{target}" ]
else:
  command = [ "push" ]  # we could keep this one as a default, and just concatenate based on the three branches above
```

I'm not a fan, but happy to change it to that, or similar, if you think it's worthwhile.